// @milkdown/plugin-prism/lib/index.js
import { findChildren } from "@milkdown/prose";
import { Plugin, PluginKey } from "@milkdown/prose/state";
import { $ctx, $prose } from "@milkdown/utils";
import { refractor } from "refractor";
import { Decoration, DecorationSet } from "@milkdown/prose/view";
//#region src/get-decorations.ts
function flatNodes(nodes, className = []) {
	return nodes.flatMap((node) => node.type === "element" ? flatNodes(node.children, [...className, ...node.properties?.className || []]) : [{
		text: node.value,
		className
	}]);
}
function getDecorations(doc, name, refractor) {
	const { highlight, listLanguages } = refractor;
	const allLanguages = listLanguages();
	const decorations = [];
	findChildren((node) => node.type.name === name)(doc).forEach((block) => {
		let from = block.pos + 1;
		const { language } = block.node.attrs;
		if (!language || !allLanguages.includes(language)) {
			console.warn("Unsupported language detected, this language has not been supported by current prism config: ", language);
			return;
		}
		flatNodes(highlight(block.node.textContent, language).children).forEach((node) => {
			const to = from + node.text.length;
			if (node.className.length) {
				const decoration = Decoration.inline(from, to, { class: node.className.join(" ") });
				decorations.push(decoration);
			}
			from = to;
		});
	});
	return DecorationSet.create(doc, decorations);
}
//#endregion
//#region src/index.ts
var prismConfig = $ctx({ configureRefractor: () => {} }, "prismConfig");
prismConfig.meta = {
	package: "@milkdown/plugin-prism",
	displayName: "Ctx<prism>"
};
var prismPlugin = $prose((ctx) => {
	const { configureRefractor } = ctx.get(prismConfig.key);
	const name = "code_block";
	return new Plugin({
		key: new PluginKey("MILKDOWN_PRISM"),
		state: {
			init: (_, { doc }) => {
				return getDecorations(doc, name, configureRefractor(refractor) ?? refractor);
			},
			apply: (transaction, decorationSet, oldState, state) => {
				const isNodeName = state.selection.$head.parent.type.name === name;
				const isPreviousNodeName = oldState.selection.$head.parent.type.name === name;
				const oldNode = findChildren((node) => node.type.name === name)(oldState.doc);
				const newNode = findChildren((node) => node.type.name === name)(state.doc);
				if (transaction.docChanged && (isNodeName || isPreviousNodeName || oldNode.length !== newNode.length || oldNode[0]?.node.attrs.language !== newNode[0]?.node.attrs.language || transaction.steps.some((step) => {
					const s = step;
					return s.from !== void 0 && s.to !== void 0 && oldNode.some((node) => {
						return node.pos >= s.from && node.pos + node.node.nodeSize <= s.to;
					});
				}))) return getDecorations(transaction.doc, name, refractor);
				return decorationSet.map(transaction.mapping, transaction.doc);
			}
		},
		props: { decorations(state) {
			return this.getState(state);
		} }
	});
});
prismPlugin.meta = {
	package: "@milkdown/plugin-prism",
	displayName: "Prose<prism>"
};
var prism = [prismPlugin, prismConfig];
//#endregion
export { prism, prismConfig, prismPlugin };
